core/container.go

```diff
@@ -896,6 +898,70 @@ func (container *Container) Publish(
 	return ref, nil
 }
 
+func (container *Container) Export(
+	ctx context.Context,
+	host *Host,
+	dest string,
+	bkClient *bkclient.Client,
+	solveOpts bkclient.SolveOpt,
+	solveCh chan<- *bkclient.SolveStatus,
+) error {
+	dest, err := host.NormalizeDest(dest)
+	if err != nil {
+		return err
+	}
+
+	payload, err := container.ID.decode()
+	if err != nil {
+		return err
+	}
+
+	st, err := payload.FSState()
+	if err != nil {
+		return err
+	}
+
+	stDef, err := st.Marshal(ctx, llb.Platform(payload.Platform))
+	if err != nil {
+		return err
+	}
+
+	cfgBytes, err := json.Marshal(specs.Image{
+		Architecture: payload.Platform.Architecture,
+		OS:           payload.Platform.OS,
+		OSVersion:    payload.Platform.OSVersion,
+		OSFeatures:   payload.Platform.OSFeatures,
+		Config:       payload.Config,
+	})
+	if err != nil {
+		return err
+	}
+
+	out, err := os.Create(dest)
```

Comment:
I guess this API is different than the local dir export one in that it directly creates files and thus forces the engine to be local to the client.

From the discussion earlier, it seems like we *might* end up going down that route anyways, but this makes the decision for us if we keep the implementation as is (unless I'm missing something of course).

My best guess at long term solution here is services (can stream this back to clients over websocket), but depending on how we end up implementing file sync between clients and remote engine maybe this will end up using that same mechanism.

Reply:
To me all these host APIs are a stopgap until we figure out how the remote engine architecture should work and how files should be synced back and forth. It's pretty hard for me to make sense of any of this API in that world.

My mental picture is that the routing API will always be co-located with Buildkit, so for this particular piece of code it shouldn't make much difference between doing an `os.Create` here vs in Buildkit against the same filesystem, it's more just dealing with the fact that his export type gives us a stream instead.

But it's strange to think about how this code would work in that context and whether it's consistent with the rest because I feel like we'd want a fundamentally different interface based on filesync anyhow. :)

Reply:
Yeah I'm okay with merging this implementation for now. The only other options would be to get the exported file into buildkit's cache somehow and then do a local export of *that*. Probably possible but not worth it.

Reply:
Sorry if I'm slow here... I want to make sure I understand.

* Directory export works by receiving the buildkit exporter tar stream and unpacking it in the target host directory;
* File export works by querying the file's contents from the buildkit state, then writing it to the target host file;
* Both Directory and File export write the result to the client's host filesystem (ie. where the client code is running)
* But @sipsma you're concerned that by using different buildkit APIs (llb state inspection vs. exporter), these two implementations may be affected differently by the future change to a remote engine architecture

I think I'm missing something as I don't understand the last part. Sorry if I'm missing something obvious.

Reply:
> * Directory export works by receiving the buildkit exporter tar stream and unpacking it in the target host directory;
> * File export works by querying the file's contents from the buildkit state, then writing it to the target host file;
> * Both Directory and File export write the result to the client's host filesystem (ie. where the client code is running)

Yes (technically it's not a tar stream but rather buildkit's own diffcopy protocol, but same idea)

> But @sipsma you're concerned that by using different buildkit APIs (llb state inspection vs. exporter), these two implementations may be affected differently by the future change to a remote engine architecture

I'm just pointing out that this API implementation is different in that nothing is streamed back to the client; we just `os.Create` the file in the engine code directly and presume that it's the same filesystem as the clients. So before we move to a remote engine we need to fix this as, unlike the other file export mechanism, it's a hard assumption on the engine being local. That's totally fine for now, I just wanted to make sure we're aware of this and tracking it (https://github.com/dagger/dagger/issues/3624)